### ISSUE TITLE
Reconstruction Error

### ISSUE THREAD

**Issue by caichuang0415** [intuitive-robots/beast_calvin#1]
Hello, thanks for your work great work on action tokenizer. I have tried it using my own action data, but the results seemed like not right.
Here's my parameter and related code:
len(joint_traj): 80
tokenizer = BSpline_Tokenizer(
    num_dof=7,                                  #  each of action has 7 freedom
    num_basis=15,                             # follow your setting in Libero dataset
    duration=len(joint_traj) / 20.0,     # 20 is the fps of my data
    seq_len=len(joint_traj),
    vocab_size=256,
    degree_p=3,
    gripper_zero_order=False,           # I choose not to handle the gripper independly
    gripper_dof=0,         
    device='cuda'
)

and below is the result after excuting the function visualize_reconstruction_error:

![Image](https://github.com/user-attachments/assets/b44b6c2b-0027-4d73-94e9-022059c2ebd8)

please tell me if this result is OK or not Ok. If not OK, what's the cause of that?

**Comment by HongyiZhouCN** [comment]
Hi caichuang,

the result does not look correct to me. The trajectory you provide looks quite smooth and 15 bases should be sufficient to fit it.

Could you try setting the duration to 1.0? that is the default setting I usually used.

Also, if possible, could you share me with 1~2 demonstrations that you are trying to fit?

Best,
Hongyi

**Comment by caichuang0415** [comment]
> Hi caichuang,
> 
> the result does not look correct to me. The trajectory you provide looks quite smooth and 15 bases should be sufficient to fit it.
> 
> Could you try setting the duration to 1.0? that is the default setting I usually used.
> 
> Also, if possible, could you share me with 1~2 demonstrations that you are trying to fit?
> 
> Best, Hongyi

Hi HongyiZhou,
Thank you for your replay. The below picture is the result after I set the Trajactory length to 20, thus the time duration is 1s (fps =20)

![Image](https://github.com/user-attachments/assets/231bce05-c7a0-4566-bd10-b4977650065e)

and this is the trajactory value before normlized ( 20 × 7):
[[0.0030412564065136483,
  0.011397065331214767,
  0.004153039984473,
  0.20769455660275007,
  0.01583192780193277,
  0.10795688293466486,
  -0.0202178955078125],
 [0.0030404060633614644,
  0.011463632408906967,
  0.004632541819576591,
  0.20772984954523133,
  0.010872839293467518,
  0.1102459393226217,
  -0.0202178955078125],
 [0.0030399455857521803,
  0.011495262690979776,
  0.004964521245309686,
  0.20773728740702876,
  0.007439597860192517,
  0.11133374251807343,
  -0.020217895507812497],
 [0.0030394721952192822,
  0.01149696531112195,
  0.0049276342347678885,
  0.2077382463575408,
  0.007821066276012478,
  0.11139399422583293,
  -0.0202178955078125],
 [0.0030195886702748765,
  0.011672412301677806,
  0.004964528327010274,
  0.2100714170168426,
  0.007439353235095415,
  0.11749334998581466,
  -0.0202178955078125],
 [0.0029248382634318813,
  0.012396147060055563,
  0.004979176806152141,
  0.22813870532704844,
  0.007285144439889449,
  0.14270762937492285,
  -0.0202178955078125],
 [0.0027553261213416535,
  0.013453141796883216,
  0.00540790514949117,
  0.25944059384325374,
  0.0028491896913942143,
  0.1796906688090619,
  -0.0202178955078125],
 [0.0025777912796007348,
  0.014606306221451794,
  0.008668525312856848,
  0.2859914205749503,
  -0.030537835368051047,
  0.22017390890612779,
  -0.0202178955078125],
 [0.0024326156316412523,
  0.015537746219212573,
  0.013297968399127114,
  0.28831794404080485,
  -0.0734406098392348,
  0.2533056340687476,
  -0.0202178955078125],
 [0.0018195291545668596,
  0.01652541514347037,
  0.019259580629448936,
  0.30029581367453656,
  -0.08167634847353575,
  0.2882291086389247,
  -0.0202178955078125],
 [0.0020166062178000738,
  0.01756942982760113,
  0.03124605959617424,
  0.30782634382646107,
  -0.052241732592717374,
  0.30849590901274454,
  -0.0202178955078125],
 [0.0034421046275634926,
  0.019123881409740574,
  0.051551754636630025,
  0.31181391845498035,
  0.014248069524644254,
  0.31678936821874953,
  -0.01983642578125],
 [0.005086894716544088,
  0.02043594806837358,
  0.06986146817578522,
  0.29504959217882487,
  0.05673653510580361,
  0.3259098710978283,
  -0.0194549560546875],
 [0.010553318155958221,
  0.02400074653186707,
  0.11001855132421846,
  0.24868097575879394,
  0.13389942198760446,
  0.3380421666524884,
  -0.0194549560546875],
 [0.014396791788601412,
  0.026696157434030688,
  0.13071094018903454,
  0.21966229697040063,
  0.20331271963151215,
  0.3485129395550562,
  -0.0194549560546875],
 [0.026761144591643495,
  0.03305475635776374,
  0.16474364778549921,
  0.16116127697863467,
  0.2962525088315681,
  0.3450516911952767,
  -0.0194549560546875],
 [0.040788309609101125,
  0.040698721506989215,
  0.19058867605006674,
  0.11436381013220927,
  0.37466079363808025,
  0.33769147455554294,
  -0.0194549560546875],
 [0.05492529521560714,
  0.050642699336754804,
  0.20668601689441007,
  0.0672495837542147,
  0.4817381191952801,
  0.321664250089693,
  -0.0194549560546875],
 [0.07018943364087527,
  0.06191335072667651,
  0.2208095876804983,
  0.016090914566066632,
  0.558878027556361,
  0.3002506589078401,
  -0.0194549560546875],
 [0.0890767742911495,
  0.07654920164520895,
  0.22998926484237925,
  -0.04314882393132983,
  0.6459622289168823,
  0.26693012091434154,
  -0.011368132108831466]]

And this is the norm parameter:
min ofr each dim: [-0.1, -0.5, -0.5, -3.0, -3.0, -3.0, -1]
max ofr each dim: [0.5, 0.5, 0.5, 3.0, 3.0, 3.0, 4.5]

Note that I did the normalization before I put the data to Beast Tokenizer

Hope for your replay, Thanks !

**Comment by HongyiZhouCN** [comment]
```
import torch
import numpy as np

from beast.models.tokenizers.bspline_tokenizer import BSpline_Tokenizer

traj = np.array([[0.0030412564065136483,
0.011397065331214767,
0.004153039984473,
0.20769455660275007,
0.01583192780193277,
0.10795688293466486,
-0.0202178955078125],
[0.0030404060633614644,
0.011463632408906967,
0.004632541819576591,
0.20772984954523133,
0.010872839293467518,
0.1102459393226217,
-0.0202178955078125],
[0.0030399455857521803,
0.011495262690979776,
0.004964521245309686,
0.20773728740702876,
0.007439597860192517,
0.11133374251807343,
-0.020217895507812497],
[0.0030394721952192822,
0.01149696531112195,
0.0049276342347678885,
0.2077382463575408,
0.007821066276012478,
0.11139399422583293,
-0.0202178955078125],
[0.0030195886702748765,
0.011672412301677806,
0.004964528327010274,
0.2100714170168426,
0.007439353235095415,
0.11749334998581466,
-0.0202178955078125],
[0.0029248382634318813,
0.012396147060055563,
0.004979176806152141,
0.22813870532704844,
0.007285144439889449,
0.14270762937492285,
-0.0202178955078125],
[0.0027553261213416535,
0.013453141796883216,
0.00540790514949117,
0.25944059384325374,
0.0028491896913942143,
0.1796906688090619,
-0.0202178955078125],
[0.0025777912796007348,
0.014606306221451794,
0.008668525312856848,
0.2859914205749503,
-0.030537835368051047,
0.22017390890612779,
-0.0202178955078125],
[0.0024326156316412523,
0.015537746219212573,
0.013297968399127114,
0.28831794404080485,
-0.0734406098392348,
0.2533056340687476,
-0.0202178955078125],
[0.0018195291545668596,
0.01652541514347037,
0.019259580629448936,
0.30029581367453656,
-0.08167634847353575,
0.2882291086389247,
-0.0202178955078125],
[0.0020166062178000738,
0.01756942982760113,
0.03124605959617424,
0.30782634382646107,
-0.052241732592717374,
0.30849590901274454,
-0.0202178955078125],
[0.0034421046275634926,
0.019123881409740574,
0.051551754636630025,
0.31181391845498035,
0.014248069524644254,
0.31678936821874953,
-0.01983642578125],
[0.005086894716544088,
0.02043594806837358,
0.06986146817578522,
0.29504959217882487,
0.05673653510580361,
0.3259098710978283,
-0.0194549560546875],
[0.010553318155958221,
0.02400074653186707,
0.11001855132421846,
0.24868097575879394,
0.13389942198760446,
0.3380421666524884,
-0.0194549560546875],
[0.014396791788601412,
0.026696157434030688,
0.13071094018903454,
0.21966229697040063,
0.20331271963151215,
0.3485129395550562,
-0.0194549560546875],
[0.026761144591643495,
0.03305475635776374,
0.16474364778549921,
0.16116127697863467,
0.2962525088315681,
0.3450516911952767,
-0.0194549560546875],
[0.040788309609101125,
0.040698721506989215,
0.19058867605006674,
0.11436381013220927,
0.37466079363808025,
0.33769147455554294,
-0.0194549560546875],
[0.05492529521560714,
0.050642699336754804,
0.20668601689441007,
0.0672495837542147,
0.4817381191952801,
0.321664250089693,
-0.0194549560546875],
[0.07018943364087527,
0.06191335072667651,
0.2208095876804983,
0.016090914566066632,
0.558878027556361,
0.3002506589078401,
-0.0194549560546875],
[0.0890767742911495,
0.07654920164520895,
0.22998926484237925,
-0.04314882393132983,
0.6459622289168823,
0.26693012091434154,
-0.011368132108831466]])


action_min = np.array([-0.1, -0.5, -0.5, -3.0, -3.0, -3.0, -1])
action_max = np.array([0.5, 0.5, 0.5, 3.0, 3.0, 3.0, 4.5])

tokenizer = BSpline_Tokenizer(num_dof=7, num_basis=10, seq_len=20, gripper_zero_order=False).to('cuda')

trajs = torch.from_numpy(traj).unsqueeze(0).to('cuda')
action_min = torch.from_numpy(action_min).unsqueeze(0).to('cuda')
action_max = torch.from_numpy(action_max).unsqueeze(0).to('cuda')

### normalize the trajectory before tokenization
trajs = (trajs - action_min) / (action_max - action_min)
### tokenize the trajectory
tokenizer.visualize_reconstruction_error(trajs)
```

**Comment by HongyiZhouCN** [comment]
Hi caichuang, could you try the code above? I test it with the trajectory you provided with the above code, it should generate result as below. 

![Image](https://github.com/user-attachments/assets/cdb6a33f-6510-4ae7-a8cd-16d794231341)

**Comment by caichuang0415** [comment]
> Hi caichuang, could you try the code above? I test it with the trajectory you provided with the above code, it should generate result as below.
> 
> ![Image](https://github.com/user-attachments/assets/cdb6a33f-6510-4ae7-a8cd-16d794231341)

Thank you very much for help me test my data. I ran your codes andf the result is the same as your result's visualiztion.

I compare my code with your code, the difference is that:  
In my code, I set the duration=len(joint_traj) / 20.0  (20 is the fps of my data), so the duration is set to 1. 
In your code, You do not pass the duration value so the duration value is set to the default value, which is the 2 * torch.pi
So the difference comes from the duration value ? I don't quite understand why you don't pass the duration value（you said the default duration for beast tokenizer is 1s. ), and why the default duration value in the code is 2 * torch.pi.
Really hope for your reply, thank you very much

![Image](https://github.com/user-attachments/assets/d35fee45-adac-4694-b365-498ec3793057)

**Comment by HongyiZhouCN** [comment]
Hi, sorry for the very late reply. We’ve uploaded a clean, lightweight implementation of BEAST to Hugging Face. In this version, we fixed the internal duration to [0, 1], since BEAST is intended to run in normalized time.
